internal/status/httproutestatus.go

```diff
@@ -0,0 +1,150 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/projectcontour/contour/internal/k8s"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	gatewayapi_v1alpha1 "sigs.k8s.io/gateway-api/apis/v1alpha1"
+)
+
+const ConditionNotImplemented gatewayapi_v1alpha1.RouteConditionType = "NotImplemented"
+const ConditionInvalid gatewayapi_v1alpha1.RouteConditionType = "Invalid"
+
+type RouteReasonType string
+
+const ReasonNotImplemented RouteReasonType = "NotImplemented"
+const ReasonPathMatchType RouteReasonType = "PathMatchType"
+const ReasonForwardTo RouteReasonType = "ForwardTo"
+const ReasonValid RouteReasonType = "Valid"
+const ReasonErrorsExist RouteReasonType = "ErrorsExist"
+
+type HTTPRouteUpdate struct {
+	FullName           types.NamespacedName
+	Conditions         []metav1.Condition
+	ExistingConditions []metav1.Condition
+	GatewayRef         types.NamespacedName
+	Generation         int64
+	TransitionTime     metav1.Time
+}
+
+// AddCondition returns a metav1.Condition for a given ConditionType.
+func (routeUpdate *HTTPRouteUpdate) AddCondition(cond gatewayapi_v1alpha1.RouteConditionType, status metav1.ConditionStatus, reason RouteReasonType, message string) metav1.Condition {
```

Comment:
This method really should do some checking by `cond` to see if there's an existing one, since conditions are supposed to be a List that's keyed of the Type. Not a blocker on this one, maybe a TODO though?

Reply:
Ohh you can't have multiple conditions of the same type? What if you had two different missing services in a ForwardTo? Or on two different paths in the same HTTPRoute?

Reply:
You'll have to concatenate the details into the Message, but yeah, the Conditions is a ListTypeMap, meaning it's a map that's implemented as a list.

Reply:
I wonder what happens if you *do* add multiple instances of the same Type, since the current code can definitely do that.

Reply:
Right now you get multiple conditions since I thought you wouldn't want to concat but you'd want to have the details. 

But like it seems with this API, my thought of how it should work is completely opposite. I'll have to come up with some concat thing to fix this since I guess I'm not following the spec in the current impl.

Reply:
Here's an example of that: https://github.com/projectcontour/contour/pull/3456/files#diff-72b53646a46f97f3db8eabaf70aa271e571f0a3e51c73a1ba593abd85d53bfebR2761-R2776

Reply:
Is this OK to merge as-is for 1.14? The current code can definitely potentially add multiple conditions of the same type, and i'm not sure what the Kube behavior is in that scenario - does it accept all of them? IMHO, given Gateway support is all alpha/WIP, I think we're probably OK as long as it doesn't cause Contour to crash or anything like that.

Reply:
added #3534 to track this